File: study-support-app/src/types/content.ts
// バックエンドの Enum に対応する型
export type BackendContentType = 'video' | 'slide' | 'pdf' | 'audio' | 'article' | 'quiz' | 'external_link';
export type ContentCategory = 'self_analysis' | 'admissions' | 'academic' | 'university_info' | 'career' | 'other';

// フロントエンドのフォーム表示などで使う可能性のある大文字版 (もし必要なら)
export type FormContentType = 'VIDEO' | 'SLIDE' | 'PDF'; // 必要に応じて他のタイプを追加

// API レスポンスに合わせた Content 型
export interface Content {
  id: string;
  title: string;
  description?: string;
  url: string;
  content_type: BackendContentType; // バックエンドの小文字 Enum 型に
  thumbnail_url?: string;
  tags?: string[];              // string[] 型に変更
  created_at: string;
  updated_at: string;
  created_by_id?: string; // APIレスポンスに含まれているか確認が必要
  average_rating?: number;
  review_count?: number;
  view_count?: number;
  category_info?: ContentCategoryInfo; // ★ 追加
  duration?: number;                   // ★ 追加
  is_premium?: boolean;                // ★ 追加
  author?: string;                     // ★ 追加
  difficulty?: number;                 // ★ 追加
  provider?: string;                   // ★ 追加 (バックエンドモデルに合わせて)
  provider_item_id?: string;           // ★ 追加 (バックエンドモデルに合わせて)
}

// 新規作成リクエスト用の型 (前回提案済み、必要であれば再確認)
export interface ContentCreate {
  title: string;
  description?: string;
  url: string;
  content_type: BackendContentType;
  thumbnail_url?: string;
  category_id?: string; // ★ Optional<UUID> に対応する型として string (または null) を許容する形に変更
  tags?: string[];
  created_by_id: string;
  // provider?: 'google' | 'slideshare' | 'speakerdeck'; // これはフォーム用の型。APIリクエストではstringでOK
  presenter_notes?: string[];
  provider?: string; // ★ 追加
  provider_item_id?: string; // ★ 追加
}

// ContentType は FormContentType として名前を変更したので古い定義は削除
// export type ContentType = 'VIDEO' | 'SLIDE' | 'PDF'; 

// ★★★ 新しいカテゴリー情報型を追加 ★★★
export interface ContentCategoryInfo {
  id: string; // UUID
  name: string;
  description?: string; // 日本語名
  display_order?: number; // ★ 追加 (元々コメントアウトされていたのを有効化)
  icon_url?: string; // ★ 追加 (元々コメントアウトされていたのを有効化)
  is_active?: boolean; // ★ 追加 (元々コメントアウトされていたのを有効化)
}

// ★★★ ContentCategoryInfo は既存で、display_order などが追加されているはず ★★★
// export interface ContentCategoryInfo {
//   id: string; 
//   name: string;
//   description?: string; 
//   display_order?: number;
//   icon_url?: string; 
//   is_active?: boolean;
// }

// --- ↓↓↓ カテゴリー管理用の新しい型定義を追加 ↓↓↓ ---
export interface ContentCategoryBase {
  name: string; 
  description?: string | null; 
  display_order?: number;
  icon_url?: string | null; 
  is_active?: boolean;
}

export type ContentCategoryCreate = ContentCategoryBase;

export interface ContentCategoryUpdate {
  name?: string;
  description?: string | null;
  display_order?: number;
  icon_url?: string | null;
  is_active?: boolean;
}
// --- ↑↑↑ カテゴリー管理用の新しい型定義を追加 ↑↑↑ --- 

File: backend/app/schemas/content.py
from pydantic import BaseModel, HttpUrl, Field, validator, computed_field
from datetime import datetime
from typing import Optional, List, Dict, Any
from uuid import UUID
from enum import Enum
from .base import TimestampMixin

class ContentType(str, Enum):
    VIDEO = "video"
    SLIDE = "slide"
    PDF = "pdf"
    AUDIO = "audio"
    ARTICLE = "article"
    QUIZ = "quiz"
    EXTERNAL_LINK = "external_link"

class ContentBase(BaseModel):
    title: str
    description: Optional[str] = None
    url: str
    content_type: ContentType
    thumbnail_url: Optional[str] = None
    duration: Optional[int] = None
    is_premium: bool = False
    author: Optional[str] = None
    difficulty: Optional[int] = Field(None, ge=1, le=5)  # 1-5の難易度

    @validator('difficulty')
    def validate_difficulty(cls, v):
        if v is not None and (v < 1 or v > 5):
            raise ValueError('難易度は1から5の間でなければなりません')
        return v

class ContentCreate(ContentBase):
    created_by_id: UUID
    tags: Optional[List[str]] = Field(default_factory=list)
    category_id: Optional[UUID] = None
    provider: Optional[str] = None
    provider_item_id: Optional[str] = None

class ContentUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    url: Optional[str] = None
    content_type: Optional[ContentType] = None
    thumbnail_url: Optional[str] = None
    category_id: Optional[UUID] = None
    duration: Optional[int] = None
    is_premium: Optional[bool] = None
    author: Optional[str] = None
    difficulty: Optional[int] = Field(None, ge=1, le=5)
    provider: Optional[str] = None
    provider_item_id: Optional[str] = None

class ContentCategoryInfo(BaseModel):
    id: UUID
    name: str  # 英語の識別子
    description: Optional[str] = None  # 日本語の表示名として利用
    display_order: Optional[int] = None
    icon_url: Optional[HttpUrl] = None
    is_active: Optional[bool] = None

    class Config:
        from_attributes = True

class ContentCategoryBase(BaseModel):
    name: str = Field(..., min_length=1, description="カテゴリの英語識別子 (例: self_analysis)")
    description: Optional[str] = Field(None, description="カテゴリの日本語名 (表示用)")
    display_order: int = Field(0, description="表示順 (小さいほど先に表示)")
    icon_url: Optional[HttpUrl] = Field(None, description="アイコン画像のURL")
    is_active: bool = Field(True, description="有効フラグ")

class ContentCategoryCreate(ContentCategoryBase):
    pass

class ContentCategoryUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1)
    description: Optional[str] = None
    display_order: Optional[int] = None
    icon_url: Optional[HttpUrl] = None
    is_active: Optional[bool] = None

class ContentResponse(ContentBase, TimestampMixin):
    id: UUID
    average_rating: Optional[float] = None
    review_count: int = 0
    view_count: int = 0
    created_by_id: Optional[UUID] = None
    category_info: Optional[ContentCategoryInfo] = None
    provider: Optional[str] = None
    provider_item_id: Optional[str] = None
    
    @computed_field
    @property
    def tags(self) -> List[str]:
        orm_tags_value = self.__dict__.get('tags')
        if isinstance(orm_tags_value, list):
            return [tag.tag_name for tag in orm_tags_value if hasattr(tag, 'tag_name')]
        return []

    class Config:
        from_attributes = True

class ContentCategoryResponse(BaseModel):
    category: ContentType
    name: str
    description: Optional[str] = None
    content_count: int
    
    class Config:
        from_attributes = True

class ReviewBase(BaseModel):
    content_id: UUID
    rating: int = Field(..., ge=1, le=5)
    comment: Optional[str] = None
    
    @validator('rating')
    def validate_rating(cls, v):
        if v < 1 or v > 5:
            raise ValueError('評価は1から5の間でなければなりません')
        return v

class ReviewCreate(ReviewBase):
    pass

class ReviewResponse(ReviewBase, TimestampMixin):
    id: UUID
    user_id: UUID
    user_name: str
    
    class Config:
        from_attributes = True

class ContentViewCreate(BaseModel):
    content_id: UUID
    watched_seconds: Optional[int] = None
    completed: bool = False

class ContentViewResponse(BaseModel):
    id: UUID
    content_id: UUID
    user_id: UUID
    watched_seconds: int
    completed: bool
    last_position: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    content: Optional[ContentResponse] = None
    
    class Config:
        from_attributes = True

class ContentRecommendationResponse(BaseModel):
    content: ContentResponse
    recommendation_reason: str
    score: float
    
    class Config:
        from_attributes = True

class FAQBase(BaseModel):
    question: str
    answer: str
    category: Optional[str] = None
    tags: Optional[List[str]] = Field(default_factory=list)

class FAQCreate(FAQBase):
    pass

class FAQResponse(FAQBase, TimestampMixin):
    id: UUID
    
    class Config:
        from_attributes = True 